package/MDAnalysis/core/universe.py

```diff
@@ -530,71 +541,140 @@ def Merge(*args):
+                attr_array.extend(attr)
+            except NameError:
+                attr_array = list(attr)
+        attrs.append(attr_class(np.array(attr_array,
+                                        dtype=attr.dtype)))
+        del attr_array
+
+    # Build up topology groups
+    for tg in (topology_groups & common_attrs):
+        bondidx = []
+        types = []
+        offset = 0
+        for ag in disjoint_atom_groups:
+            bonds = getattr(ag, tg)
+            bond_class = type(getattr(ag.universe._topology, tg))
+            bondidx.extend(bonds.indices + offset)
+            if hasattr(bonds, '_bondtypes'):
+                types.extend(bonds.types())
+            else:
+                types.extend([None]*len(bonds))
+            offset += len(ag)
+        bondidx = np.array(bondidx, dtype=np.int32)
+        if any(t is None for t in types):
+            attrs.append(bond_class(values))
+        else:
+            types = np.array(types, dtype='|S8')
+            attrs.append(bond_class(bondidx, types))
+
+    # Renumber residue and segment indices
+    n_atoms = sum([len(ag) for ag in disjoint_atom_groups])
+    residx = []
+    segidx = []
+    for ag in disjoint_atom_groups:
+        res_offset = len(set(residx))
+        resdict = {n: i+res_offset for i, n in enumerate(set(ag.resindices))}
+        seg_offset = len(set(segidx))
+        segdict = {n: i+len(set(segidx)) for i, n in enumerate(set(
+                                                     ag.segindices))}
+        residx.extend([resdict[n] for n in ag.resindices])
+        segidx.extend([segdict[n] for n in ag.segindices])
+
+    residx = np.array(residx, dtype=np.int32)
+    segidx = np.array(segidx, dtype=np.int32)
+
+    n_residues = len(set(residx))
+    n_segments = len(set(segidx))
+
+    top = Topology(n_atoms, n_residues, n_segments,
+                   attrs=attrs,
+                   atom_resindex=residx,
+                   residue_segindex=segidx)
+
+    # Create blank Universe and put topology in it
+    u = Universe()
+    u._topology = top
+
+    # generate Universe version of each class
+    # AG, RG, SG, A, R, S
+    u._classes = groups.make_classes()
+
+    # Put Group level stuff from topology into class
+    for attr in u._topology.attrs:
+        u._process_attr(attr)
+
+    # Generate atoms, residues and segments
+    u.atoms = u._classes['atomgroup'](
+            np.arange(u._topology.n_atoms), u)
+
+    u.residues = u._classes['residuegroup'](
+            np.arange( u._topology.n_residues), u)
+
+    u.segments = u._classes['segmentgroup'](np.arange(
+        u._topology.n_segments), u)
+
+    # Update Universe namespace with segids
+    for seg in u.segments:
+        if hasattr(seg, 'segid'):
+            if seg.segid[0].isdigit():
+                name = 's' + seg.segid
+            else:
+                name = seg.segid
+            u.__dict__[name] = seg
+
+    coords = np.vstack([a.positions for a in disjoint_atom_groups])
```

Comment:
No - I was a bit silly implementing this the way I did, shouldn't be changing the behavior from the old Merge. I was trying to avoid thinking about how residue indices should be assigned to atoms in the following merges, if universes `u` and `u2` each have 10 atoms belonging to 1 residue:

    mda.Merge(u.atoms[:5], u.atoms[5:10])
    mda.Merge(u.atoms, u.atoms)
    mda.Merge(u.atoms, u2.atoms, u.atoms)

and tried avoiding the problem by adding together atomgroups from the same universe first, but I disrupted the order in the process.

So question - if atomgroups from the same universe belong to the same residue/segment before the merge, should they have the same residue/segment after the merge, or they should have different residue/segment indices after the merge?

I'm taking for granted that residues from different universes before the merge should not belong to the same residues (i.e. have the same residue indices) after the merge.